src/accelerate/utils/modeling.py

```diff
@@ -302,6 +307,24 @@ def set_module_tensor_to_device(
     is_buffer = tensor_name in module._buffers
     old_value = getattr(module, tensor_name)
 
+    # Treat the case where old_value (or a custom `value`, typically offloaded to RAM/disk) belongs to a tied group, and one of the weight
+    # in the tied group has already been dispatched to the device, by avoiding reallocating memory on the device and just copying the pointer.
+    if (
+        value is not None
+        and tied_params_map is not None
+        and value.data_ptr() in tied_params_map
+        and device in tied_params_map[value.data_ptr()]
+    ):
+        module._parameters[tensor_name] = tied_params_map[value.data_ptr()][device]
+        return
```

Comment:
We need to treat the case `value is not None` separately. This is because it corresponds to the offloading case, where (at least for RAM offloading) `value` from `self.weights_map` is indeed shared for tied weights.

When we move `self.weights_map[name]` to device, we want to do it only once. Especially, given that we may have nested modules that refer to the same data (see `test_dispatch_model_tied_weights_memory_with_nested_offload` for an example), we may have several `pre_forward` calls that use the same data, so we need that here.